src/OpenTelemetry.Instrumentation.EventCounters/EventCounterListener.cs

```diff
@@ -0,0 +1,245 @@
+
+        private enum InstrumentType
+        {
+            Gauge,
+            Counter,
+        }
+
+        protected override void OnEventSourceCreated(EventSource source)
+        {
+            // TODO: Add Configuration options to selectively subscribe to EventCounters
+
+            try
+            {
+                var arguments = new Dictionary<string, string>
+                {
+                    ["EventCounterIntervalSec"] = EventCounterMetricsOptions.RefreshIntervalSecs.ToString(),
+                };
+
+                this.EnableEvents(source, EventLevel.Verbose, EventKeywords.All, arguments);
+            }
+            catch (Exception ex)
+            {
+                EventCountersInstrumentationEventSource.Log.ErrorEventCounter(source.Name, ex.Message);
+            }
+        }
+
+        protected override void OnEventWritten(EventWrittenEventArgs eventData)
+        {
+            if (!this.isInitialized || !eventData.EventName.Equals("EventCounters"))
+            {
+                return;
+            }
+
+            try
+            {
+                if (eventData.Payload.Count > 0 && eventData.Payload[0] is IDictionary<string, object> eventPayload)
+                {
+                    this.ExtractAndPostMetric(eventData.EventSource.Name, eventPayload);
+                }
+                else
+                {
+                    EventCountersInstrumentationEventSource.Log.IgnoreEventWrittenAsEventPayloadNotParseable(eventData.EventSource.Name);
+                }
+            }
+            catch (Exception ex)
+            {
+                EventCountersInstrumentationEventSource.Log.ErrorEventCounter(eventData.EventName, ex.ToString());
+            }
+        }
+
+        private void ExtractAndPostMetric(string eventSourceName, IDictionary<string, object> eventPayload)
+        {
+            try
+            {
+                bool calculateRate = false;
+                double actualValue = 0.0;
+                double actualInterval = 0.0;
+                double recordedValue = 0.0;
+
+                string counterName = string.Empty;
+                string counterDisplayName = string.Empty;
+                InstrumentType instrumentType = InstrumentType.Counter;
+
+                foreach (KeyValuePair<string, object> payload in eventPayload)
+                {
+                    var key = payload.Key;
+                    if (key.Equals("Name", StringComparison.OrdinalIgnoreCase))
+                    {
+                        counterName = payload.Value.ToString();
+                    }
+                    else
+                    if (key.Equals("DisplayName", StringComparison.OrdinalIgnoreCase))
+                    {
+                        counterDisplayName = payload.Value.ToString();
+                    }
+                    else if (key.Equals("Mean", StringComparison.OrdinalIgnoreCase))
+                    {
+                        instrumentType = InstrumentType.Counter;
```

Comment:
not reading the value anywhere?

Reply:
instrument type is passed to RecordMetric in line 162.

Reply:
but where do we read the actual value from the payload?

Reply:
We are inferring it. If countertype field is mean, then the mean field is available. If it is Sum, we look at the increment field. So it is effectively the same , but I can clarify the intent by reading the field. Here is some diagnostics code I am basing my thinking on: https://github.com/dotnet/diagnostics/blob/9ccbd1e94c1571c4af35bda26e580474aed3a37f/src/Microsoft.Diagnostics.Monitoring.EventPipe/Counters/TraceEventExtensions.cs#L41

Reply:
I am not sure I follow. If we are not reading the value from the payload here, we wont get the value from anywhere else. Even the linked code is reading the value from payload.

Reply:
Ok yes thats a problem - fixing it.